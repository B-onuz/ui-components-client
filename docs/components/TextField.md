# TextField (Component)

[Descrição do componente e onde usar]

## Como utilizar

```js
import TextField from 'ui-components/lib/components/TextField'
```

```jsx
import React, { useState } from 'react'

const FooBar = () => {
  const [firstname, setFirstName] = useState('')
  return (
    <TextField
      id="name"
      name={'name'}
      required
      label={'Nome'}
      placeholder={'Nome'}
      value={firstname}
      onChange={(e) => setFirstName(e.target.value)}
    />
  )
}

export default FooBar
```

## Variações

### Multiline

```jsx
import React, { useState } from 'react'

const FooBar = () => {
  const [firstname, setFirstName] = useState('')
  return (
    <TextField
      id="name"
      name={'name'}
      required
      multiline // para funcionar como textarea, basta adicionar esta prop
      label={'Nome'}
      placeholder={'Nome'}
      value={firstname}
      onChange={(e) => setFirstName(e.target.value)}
    />
  )
}

export default FooBar
```

## Styled-system props

- space https://styled-system.com/table#space
- layout https://styled-system.com/table#layout
