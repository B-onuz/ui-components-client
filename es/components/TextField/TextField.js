"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Label = _interopRequireDefault(require("../Label"));

var _Input = _interopRequireDefault(require("../Input"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledSystem = require("styled-system");

var _propTypes2 = _interopRequireDefault(require("@styled-system/prop-types"));

var _Textarea = _interopRequireDefault(require("../Textarea"));

var _FormHelperText = _interopRequireDefault(require("../FormHelperText"));

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/TextField/TextField.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  display: block;\n  font-family: inherit;\n  ", "\n  label {\n    display: block;\n    margin-bottom: 0.4em;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledTextField = _styledComponents["default"].div(_templateObject(), (0, _styledSystem.compose)(_styledSystem.space, _styledSystem.layout));

var TextField = function TextField(_ref) {
  var children = _ref.children,
      label = _ref.label,
      placeholder = _ref.placeholder,
      name = _ref.name,
      id = _ref.id,
      value = _ref.value,
      onChange = _ref.onChange,
      required = _ref.required,
      error = _ref.error,
      _ref$inputProps = _ref.inputProps,
      inputProps = _ref$inputProps === void 0 ? {} : _ref$inputProps,
      multiline = _ref.multiline,
      _ref$helperText = _ref.helperText,
      helperText = _ref$helperText === void 0 ? '' : _ref$helperText,
      _ref$type = _ref.type,
      type = _ref$type === void 0 ? 'text' : _ref$type,
      rest = _objectWithoutPropertiesLoose(_ref, ["children", "label", "placeholder", "name", "id", "value", "onChange", "required", "error", "inputProps", "multiline", "helperText", "type"]);

  var InputComponent = multiline ? _Textarea["default"] : _Input["default"];
  return /*#__PURE__*/_react["default"].createElement(StyledTextField, _extends({}, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 39,
      columnNumber: 5
    }
  }), /*#__PURE__*/_react["default"].createElement(_Label["default"], {
    htmlFor: id,
    color: !!error && 'danger',
    required: !!required,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 40,
      columnNumber: 7
    }
  }, label), /*#__PURE__*/_react["default"].createElement(InputComponent, _extends({
    id: id,
    name: name,
    placeholder: placeholder
    /* controlled props */
    ,
    value: value,
    onChange: onChange
    /* validation props */
    ,
    required: !!required,
    "aria-required": !!required,
    "aria-invalid": !!error,
    type: type,
    error: error
  }, inputProps, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 43,
      columnNumber: 7
    }
  })), !!error && /*#__PURE__*/_react["default"].createElement(_FormHelperText["default"], {
    color: 'red',
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 58,
      columnNumber: 19
    }
  }, error), !!helperText && /*#__PURE__*/_react["default"].createElement(_FormHelperText["default"], {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 59,
      columnNumber: 24
    }
  }, helperText));
};

TextField.propTypes = process.env.NODE_ENV !== "production" ? _extends({}, _propTypes2["default"].space, {}, _propTypes2["default"].layout, {
  multiline: _propTypes["default"].bool
}) : {};
TextField.defaultProps = {
  required: false,
  error: false,
  multiline: false
};
var _default = TextField;
exports["default"] = _default;