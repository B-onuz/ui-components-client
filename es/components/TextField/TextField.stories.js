'use strict'

Object.defineProperty(exports, '__esModule', {
  value: true,
})
exports.TextFieldMultiline = exports.TextFieldExample = exports['default'] = void 0

var _react = _interopRequireDefault(require('react'))

var _TextField = _interopRequireDefault(require('./TextField'))

var _reflexbox = require('reflexbox')

var _this = void 0,
  _jsxFileName = '/home/junior/Fontes/ui-components-client/src/components/TextField/TextField.stories.js'

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj }
}

var _default = {
  title: 'TextField',
  component: _TextField['default'],
}
exports['default'] = _default

var TextFieldExample = function TextFieldExample() {
  return /*#__PURE__*/ _react['default'].createElement(
    _reflexbox.Flex,
    {
      flexWrap: 'wrap',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 11,
        columnNumber: 3,
      },
    },
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 12,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'name',
        name: 'name',
        required: true,
        label: 'Nome',
        m: 2,
        placeholder: 'Nome',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 13,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 15,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'lastname',
        name: 'lastname',
        required: true,
        label: 'Sobrenome',
        m: 2,
        placeholder: 'Sobrenome',
        error: 'Este campo é obrigatório.',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 16,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 18,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'fullname',
        name: 'fullname',
        label: 'Exibir Como:',
        m: 2,
        placeholder: 'Nome + Sobrenome',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 19,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 21,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'fullname',
        name: 'fullname',
        label: 'Exibir Como:',
        m: 2,
        placeholder: 'Nome + Sobrenome',
        helperText: 'Exemplo: John Doe',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 22,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 24,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'email',
        type: 'email',
        name: 'email',
        label: 'Email',
        m: 2,
        placeholder: 'Email',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 25,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 27,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'telefone',
        type: 'phone',
        name: 'fullname',
        label: 'Telefone',
        m: 2,
        placeholder: 'Telefone',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 28,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 30,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'fullname',
        name: 'fullname',
        label: 'Exibir Como:',
        m: 2,
        placeholder: 'Nome + Sobrenome',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 31,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 8],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 33,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'number',
        type: 'number',
        name: 'number',
        label: 'Número',
        m: 2,
        placeholder: 'Número',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 34,
          columnNumber: 7,
        },
      })
    )
  )
}

exports.TextFieldExample = TextFieldExample

var TextFieldMultiline = function TextFieldMultiline() {
  return /*#__PURE__*/ _react['default'].createElement(
    _reflexbox.Flex,
    {
      flexWrap: 'wrap',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 40,
        columnNumber: 3,
      },
    },
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 41,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'comment',
        name: 'comment',
        multiline: true,
        required: true,
        label: 'Comentário',
        m: 2,
        placeholder: 'Comentário',
        inputProps: {
          maxLength: 100,
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 42,
          columnNumber: 7,
        },
      })
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _reflexbox.Box,
      {
        width: [1, 1 / 2, 1 / 3],
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 53,
          columnNumber: 5,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(_TextField['default'], {
        id: 'comment',
        error: 'Error message here',
        name: 'comment',
        multiline: true,
        required: true,
        label: 'Comentário',
        m: 2,
        placeholder: 'Comentário',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 54,
          columnNumber: 7,
        },
      })
    )
  )
}

exports.TextFieldMultiline = TextFieldMultiline
