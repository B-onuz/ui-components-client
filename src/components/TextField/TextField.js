import React from 'react'
import styled from 'styled-components'
import Label from '../Label'
import Input from '../Input'
import PropTypes from 'prop-types'
import { space, layout, compose, flexbox } from 'styled-system'
import propTypes from '@styled-system/prop-types'
import Textarea from '../Textarea'
import FormHelperText from '../FormHelperText'

const StyledTextField = styled.div`
  display: block;
  font-family: inherit;
  ${compose(space, layout)}
  label {
    display: block;
    margin-bottom: 0.4em;
  }
`

const TextField = ({
  children,
  label,
  placeholder,
  name,
  id,
  value,
  onChange,
  required,
  error,
  inputProps = {},
  multiline,
  helperText = '',
  type = 'text',
  ...rest
}) => {
  const InputComponent = multiline ? Textarea : Input
  return (
    <StyledTextField {...rest}>
      <Label htmlFor={id} color={!!error && 'danger'} required={!!required}>
        {label}
      </Label>
      <InputComponent
        id={id}
        name={name}
        placeholder={placeholder}
        /* controlled props */
        value={value}
        onChange={onChange}
        /* validation props */
        required={!!required}
        aria-required={!!required}
        aria-invalid={!!error}
        type={type}
        error={error}
        {...inputProps}
      />
      {!!error && <FormHelperText color={'red'}>{error}</FormHelperText>}
      {!!helperText && <FormHelperText>{helperText}</FormHelperText>}
    </StyledTextField>
  )
}

TextField.propTypes = {
  ...propTypes.space,
  ...propTypes.layout,
  multiline: PropTypes.bool,
}

TextField.defaultProps = {
  required: false,
  error: false,
  multiline: false,
}

export default TextField
