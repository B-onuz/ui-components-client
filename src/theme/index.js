const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
