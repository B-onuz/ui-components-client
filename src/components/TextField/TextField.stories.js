import React from 'react'
import TextField from './TextField'
import { Box, Flex } from 'reflexbox'

export default {
  title: 'TextField',
  component: TextField,
}

export const TextFieldExample = () => (
  <Flex flexWrap={'wrap'}>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="name" name={'name'} required label={'Nome'} m={2} placeholder={'Nome'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="lastname" name={'lastname'} required label={'Sobrenome'} m={2} placeholder={'Sobrenome'} error={'Este campo é obrigatório.'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="fullname" name={'fullname'} label={'Exibir Como:'} m={2} placeholder={'Nome + Sobrenome'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="fullname" name={'fullname'} label={'Exibir Como:'} m={2} placeholder={'Nome + Sobrenome'} helperText={'Exemplo: John Doe'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="email" type="email" name={'email'} label={'Email'} m={2} placeholder={'Email'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="telefone" type="phone" name={'fullname'} label={'Telefone'} m={2} placeholder={'Telefone'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField id="fullname" name={'fullname'} label={'Exibir Como:'} m={2} placeholder={'Nome + Sobrenome'} />
    </Box>
    <Box width={[1, 1 / 2, 1 / 8]}>
      <TextField id="number" type={'number'} name={'number'} label={'Número'} m={2} placeholder={'Número'} />
    </Box>
  </Flex>
)

export const TextFieldMultiline = () => (
  <Flex flexWrap={'wrap'}>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField
        id="comment"
        name={'comment'}
        multiline
        required
        label={'Comentário'}
        m={2}
        placeholder={'Comentário'}
        inputProps={{ maxLength: 100 }}
      />
    </Box>
    <Box width={[1, 1 / 2, 1 / 3]}>
      <TextField
        id="comment"
        error={'Error message here'}
        name={'comment'}
        multiline
        required
        label={'Comentário'}
        m={2}
        placeholder={'Comentário'}
      />
    </Box>
  </Flex>
)
