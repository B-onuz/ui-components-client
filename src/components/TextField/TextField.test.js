import React from 'react'
import 'jest-styled-components'
import TextField from './TextField.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'

test('TextField changes the class when hovered', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <TextField id="name" name={'name'} required label={'Nome'} m={2} placeholder={'Nome'} />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
