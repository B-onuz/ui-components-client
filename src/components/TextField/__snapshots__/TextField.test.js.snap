// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField changes the class when hovered 1`] = `
.c1 {
  color: #555;
  font-size: 0.9em;
  font-weight: 600;
}

.c2 {
  border: 1px solid #dededf;
  border-radius: 4px;
  line-height: 2.1em;
  outline: none;
  padding-right: 1em;
  padding-left: 1em;
  width: 100%;
  box-sizing: border-box;
  font-size: 16px;
}

.c2::-webkit-input-placeholder {
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  color: #b5b5b5;
}

.c2::-moz-placeholder {
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  color: #b5b5b5;
}

.c2:-ms-input-placeholder {
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  color: #b5b5b5;
}

.c2::placeholder {
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  color: #b5b5b5;
}

.c2:focus {
  box-shadow: 0 0 3px rgba(33,133,208,0.5);
  border-color: #2185d0;
}

.c0 {
  display: block;
  font-family: inherit;
  margin: 8px;
}

.c0 label {
  display: block;
  margin-bottom: 0.4em;
}

<div
  className="c0"
>
  <label
    className="c1"
    color={false}
    htmlFor="name"
  >
    <abbr
      title="Campo obrigatÃ³rio"
    >
      *
    </abbr>
     Nome
  </label>
  <input
    aria-invalid={false}
    aria-required={true}
    className="c2"
    color="blue"
    id="name"
    name="name"
    placeholder="Nome"
    required={true}
    size="medium"
    type="text"
  />
</div>
`;
